src/app/clusters/thread-border-router-management-server/thread-border-router-management-server.cpp

```diff
@@ -0,0 +1,384 @@
+        }
+        break;
+    }
+    case Attributes::BorderRouterName::Id: {
+        char borderRouterNameBuf[kBorderRouterNameMaxLength + 1];
+        MutableCharSpan borderRouterName(borderRouterNameBuf);
+        status = ReadBorderRouterName(borderRouterName);
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(borderRouterName);
+        }
+        break;
+    }
+    case Attributes::BorderAgentID::Id: {
+        uint8_t borderAgentIDBuf[kBorderAgentIdLength];
+        MutableByteSpan borderAgentID(borderAgentIDBuf);
+        status = ReadBorderAgentID(borderAgentID);
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(borderAgentID);
+        }
+        break;
+    }
+    case Attributes::ThreadVersion::Id: {
+        uint16_t threadVersion;
+        status = ReadThreadVersion(threadVersion);
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(threadVersion);
+        }
+        break;
+    }
+    case Attributes::InterfaceEnabled::Id: {
+        bool interfaceEnabled;
+        status = ReadInterfaceEnabled(interfaceEnabled);
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(interfaceEnabled);
+        }
+        break;
+    }
+    case Attributes::ActiveDatasetTimestamp::Id: {
+        Optional<uint64_t> activeDatasetTimestamp;
+        status = ReadActiveDatasetTimestamp(activeDatasetTimestamp);
+        if (status == CHIP_NO_ERROR)
+        {
+            status = activeDatasetTimestamp.HasValue() ? aEncoder.Encode(DataModel::MakeNullable(activeDatasetTimestamp.Value()))
+                                                       : aEncoder.EncodeNull();
+        }
+        break;
+    }
+    default:
+        break;
+    }
+    return status;
+}
+
+void ServerInstance::CommitSavedBreadcrumb()
+{
+    if (mBreadcrumb.HasValue())
+    {
+        GeneralCommissioning::SetBreadcrumb(mBreadcrumb.Value());
+    }
+    mBreadcrumb.ClearValue();
+}
+
+void ServerInstance::OnActivateDatasetComplete(uint32_t sequenceNum, CHIP_ERROR error)
+{
+    auto commandHandleRef = std::move(mAsyncCommandHandle);
+    auto commandHandle    = commandHandleRef.Get();
+    if (commandHandle == nullptr)
+    {
+        return;
+    }
+    if (mSetActiveDatasetSequenceNumber != sequenceNum)
+    {
+        // Previous SetActiveDatasetRequest was handled.
+        return;
+    }
+    if (error == CHIP_NO_ERROR)
+    {
+        // The successful completion of the activation process SHALL disarm the fail-safe timer.
+        DisarmFailSafeTimer();
+        CommitSavedBreadcrumb();
+    }
+    else
+    {
+        ChipLogError(Zcl, "Failed on activating the active dataset for Thread BR: %" CHIP_ERROR_FORMAT, error.Format());
+    }
+    commandHandle->AddStatus(mPath, StatusIB(error).mStatus);
+}
+
+void ServerInstance::OnFailSafeTimerExpired()
+{
+    if (mDelegate)
+    {
+        mDelegate->RevertActiveDataset();
```

Comment:
Does this revert take care of dealing with the situation where there was previously an active dataset and it got changes?

Reply:
The active dataset can be set only if there was no previous active dataset configured so we don't need to restore anything. On `RevertActiveDataset` I think we only need to clear the active dataset from the Thread stack so that a new `SetActiveDataset` operation is possible in the future.  

This is already implemented in `RevertActiveDataset` and the functionality looks fine in my view.

Reply:
Since the `SetActiveDataset` will only be called when there is no active dataset configured, so this revert function will do the following things: 1. Check whether there is a previous `SetActiveDataset` called, 2. If yes, clear the active dataset stored by the Thread stack.